Jsonrpc20/wip

brbzull0 opened the pull request:
**WIP, The whole idea of this pr is to get internal feedback on the direction this is taking, and allow comments, base on this I will take the appropiate action**

_few notes_

- I've placed TODO's to keep track on some things that needs to be done.
- Some of the code is not yet PR quality
- There is not much documentation in place yet.
- **I've created a `mgmt2` folder to start clean and avoid  issues in the future when removing 'traffic_manager**', there is some code like `FileManager.cc/h` which was copy/paste from TM with some mods in order to be reused. Writing them both again is a possibility I'm considering.
- There are some coupled modules that needs dependencies from the current `mgmt` folder which should be removed in the future.
- RPC server message validation is done by parsing the incoming message to check if it's a valid json, it does also have a limit on the buffer, it's doing both things right now, **this will change**.
- RPC server is blocking.
- JSONRPC protocol is not 100% honor as the parsing library(yamlcpp) have some limitation.
- Some of the validity check when setting a new config variable was copy/paste (like 'recordValidityCheck'), there are better ways to achieve this, this code may change in the future.

Some todos:

- [ ]  detect the incoming message format, either `yaml` or `json` and asnwer in the same way. (Node::Style) should give us this information

Commit:
JSONRPC 2.0:
- Add JSONRPC 2.0 protocol implementation.
- Add Unix Domain Socket Server implementation hooked to the jsonrpc impl.
- Unit tests.
- Add public handlers for configuration commands.
- Add public handler for metric commands.
- Add storage handler to set a device off.
- Remove include folder from Makefile. Files now uses proper path
- Add basic server start/stop drain handler
- Add some helper functions to deal with errors.
- Add utility file to be shared among the rpc handlers.
- Remove null(yaml ~) id from responses that should send back json null field.
- Use utils helpers class
- Tidy up  some code.
- Add YAML converter for RecUpdateT enum
- Move utility functions to a common file.
- General cleanup on some large functions.
- Add public handler for config reload.
- Copy FileManager from mgmt to mgmt2 and make it Singleton.
- Move response structure into the jsonrpc private header.
- Add public handler to manage host(up, down) commands.
- Add public handler to support basic plugin messages.

Commit:
Errata:
- Implemen Rv constructor that takes an rvalue Errata.
- Implement Errata constructor to take an std::error_code
- Implenent Errata::push(Errata) method (was declared).

Commit:
JSONRPC:
- Configuration Handler, replace std::pair for a names structure.
- Rename ErrorId to ErrorUtils to share error codes boundaries among different handlers.
- Use proper Error Code.
- Avoid failing on first error and instead, append all errors and no erros to the result. This can be used by the JSONRP to return both errors and valid response content.

Commit:
JSONRPC:
-Add public handler for shutdown and restart.

SolidWallOfCode commented on mgmt2/rpc/common/JsonRpcApi.h:
```diff
@@ -0,0 +1,32 @@
+
+#include "tscore/Errata.h"
+
+namespace rpc::jsonrpc
```
Is the doubled "rpc" needed?

SolidWallOfCode commented on mgmt2/rpc/config/JsonRpcConfig.h:
```diff
@@ -0,0 +1,74 @@
+
+namespace rpc::config
+{
+template <class... Ts> struct overloaded : Ts... {
```
Should look at moving this to "ts_meta.h"

brbzull0 replied:
This actually needs to be removed, as I did  drop the `std::variant`. If I use it again, I'll move this to `ts_meta.h`

SolidWallOfCode commented on mgmt2/rpc/config/JsonRpcConfig.h:
```diff
@@ -0,0 +1,74 @@
+    }
+  }
+
+  // this will throw.
```
Under what circumstances? If `_transportConfig` is default constructed?

SolidWallOfCode commented on mgmt2/rpc/handlers/Admin.h:
```diff
@@ -0,0 +1,75 @@
+namespace rpc::admin
+{
+void
+register_admin_jsonrpc_handlers()
```
This shouldn't be in a header file.

brbzull0 replied:
Ok

SolidWallOfCode commented on mgmt2/rpc/handlers/common/RecordsUtils.cc:
```diff
@@ -0,0 +1,313 @@
+}
+
+const RPCRecordErrorCategory &
+get_record_error_category()
```
Why is this better wrapped, instead of using a file scope value?

brbzull0 replied:
Actually , this function should be in the header(publicly exposed) file so you can eventually compare the error category when just dealing with the `std::error_code` API:
```
void handle_error(std::error_code const& ec) {
   if (ec.category() == get_record_error_category()) {
     ....
   } else if (...) {
   ...
  }
}
```

SolidWallOfCode commented on mgmt2/rpc/handlers/common/RecordsUtils.cc:
```diff
@@ -0,0 +1,313 @@
+      int r = pcre_exec(regex, nullptr, value.data(), value.size(), 0, 0, nullptr, 0);
+
+      pcre_free(regex);
+      return (r != -1) ? true : false;
```
`return r != -1;` ?

SolidWallOfCode commented on mgmt2/rpc/handlers/common/RecordsUtils.cc:
```diff
@@ -0,0 +1,313 @@
+  recordRangeCheck(std::string_view pattern, std::string_view value)
+  {
+    char *p = const_cast<char *>(pattern.data());
+    Tokenizer dashTok("-");
```
Don't use `Tokenizer`, it's junk. Use `TextView`.

SolidWallOfCode commented on mgmt2/rpc/handlers/common/RecordsUtils.cc:
```diff
@@ -0,0 +1,313 @@
+    //  regex_t regex;
+    //  int result;
+    bool check;
+    std::string_view range_pattern = R"(\[[0-9]+\-[0-9]+\]\\\.\[[0-9]+\-[0-9]+\]\\\.\[[0-9]+\-[0-9]+\]\\\.\[[0-9]+\-[0-9]+\])";
```
You should either use existing support in "ink_inet.h" or add this if it's not there. One change that could be done is to enable the address parsing to take a `nullptr` for `sockaddr*` to indicate only parsing is to be done, not storing, for exactly this kind of validity check.

brbzull0 replied:
most of the code in `RecordsUtils.cc` which is used for validation(just for `config set` is from TM -> https://github.com/brbzull0/trafficserver/blob/25d13c1b33117d1969971fa4c10b55077438d880/mgmt/WebMgmtUtils.cc#L1063-L1071

I may review this and see if we can use a different(or existing) approach.

SolidWallOfCode commented on mgmt2/rpc/handlers/common/convert.h:
```diff
@@ -0,0 +1,121 @@
+
+namespace record::field_names
+{
+static constexpr auto Name{"name"};
```
I think these should be `std::string` - AFAICT if not, then the value will be converted to `std::string` for comparison, making for a lot of unnecessary allocations.

brbzull0 replied:
It should call the right overload for `operator==`
```
template< class CharT, class Traits, class Alloc >
bool operator==( const std::basic_string<CharT,Traits,Alloc>& lhs,
                 const CharT* rhs );
```
I'll make sure this is the case.

SolidWallOfCode commented on mgmt2/rpc/handlers/common/convert.h:
```diff
@@ -0,0 +1,121 @@
+      node[field::SyntaxCheck]  = record.config_meta.check_expr ? record.config_meta.check_expr : "null";
+      node[field::RecordClass]  = static_cast<int>(record.rec_type);
+      if (record.name) {
+        if (const auto it = ts::Overridable_Txn_Vars.find(record.name); it != ts::Overridable_Txn_Vars.end()) {
```
Maybe
```
const auto it = ts::Overridable_Txn_Vars.find(record.name);
node[field::Overridable] = ( it == ts::Overridable_Txn_Vars.end()) ? "false" : "true";
```

brbzull0 replied:
absolutely yes.

SolidWallOfCode commented on mgmt2/rpc/jsonrpc/Defs.h:
```diff
@@ -0,0 +1,122 @@
+struct RpcRequestInfo {
+  std::string jsonrpc;
+  std::string method;
+  std::optional<std::string> id;
```
An empty string isn't sufficient?

brbzull0 replied:
I think this express the meaning of the `id` (now the question could be to whom? as it's not a public API.) I'll think about that.

SolidWallOfCode commented on mgmt2/rpc/jsonrpc/JsonRpc.cc:
```diff
@@ -0,0 +1,261 @@
+    return invoke_handler(request);
+  }
+
+  return response_type{};
```
Or `return {};`

SolidWallOfCode commented on mgmt2/rpc/jsonrpc/JsonRpc.cc:
```diff
@@ -0,0 +1,261 @@
+  std::lock_guard<std::mutex> lock(_mutex);
+
+  const auto found = std::find_if(std::begin(_methods), std::end(_methods), [&](auto const &p) { return p.first == methodName; });
+  return found != std::end(_methods) ? std::make_pair(found->second, true) : std::make_pair(nullptr, false);
```
`nullptr` doesn't suffice as a failure marker?

brbzull0 replied:
it should.

Commit:
JSONRPC:
- Split public handler registation into a h,cc file. Also move the files into ts source code folder.

brbzull0 commented on mgmt2/rpc/server/JsonRpcConfig.h:
```diff
@@ -0,0 +1,74 @@
+/**
```
Ignore this file. it's a leftover from a previous folder.

Commit:
JSONRPC:
- Remove id from the Errata ctor when taking a std::error_code.
- Use the proper Errata ctor that takes a ec.

Commit:
JSONRPC 2.0
- Rename file that holds the registration functions to be exposed as public handlers.
- Configuration support.
 - Add configuration file support for jsonrpc server, including the transport one(UDS)
 - Add Comments on some of the configuration classes.
 - Add unit test for the new configuration file reading.

Commit:
JSONRPC - Add code documentation around JsonRpc and the YAML Codec.

Commit:
JSONRPC - Rename transport classes and use Communication instead. This change touches the server and the configuration.

Commit:
JSONRPC:
- Add generic record lookup function.
- remove specific metric and config record lookup function.
- remove metric files that are not longer needed as they were only to lookup for metrics

Commit:
JSONRPC: Add basic documentation.

Commit:
JSONRPC: unittest, moving code from the rpc test client to the test file.

brbzull0 commented on mgmt2/rpc/handlers/common/convert.h:
```diff
@@ -0,0 +1,175 @@
+  static Node
+  encode(const RecConfigMeta &configMeta)
+  {
+    Node node;
```
try catch missing?

brbzull0 commented on mgmt2/rpc/handlers/common/convert.h:
```diff
@@ -0,0 +1,175 @@
+    } catch (std::exception const &e) {
+      // we create an empty map node, we do not want to have a null. revisit this.
+      YAML::NodeType::value kind = YAML::NodeType::Map;
+      node                       = YAML::Node{kind};
```
two lines?

Commit:
JSONRPC 2.0
- Name changes around the server, jsonrpc protocol, errors, namespaces.
- Improve docs.
- fix some issues

Commit:
JSONRPC 2.0
- Add support for traffic_ctl - new traffic_ctl binary that will talk
  jsonrpc.
- Add documentation for traffic_ctl, jsonrpc handlers.
- Fix and improve some jsonrpc code.(server, protocol)
- Add configure option to build either legacy or new traffic_ctl

Commit:
JSONRPC 2.0
- Fix build - work-around for modern automake
- Fix ubuntu ci build. Fix some doc issues.
- Fix documentation.  Remove comments.
- traffic_ctl: Add better errors when connecting with the TS socket.

Commit:
JSONRPC 2.0.
traffic_ctl - An invalid remote method call was done during config set.

Commit:
JSONRPC 2.0: Fix make distclean issue.

Commit:
JSONRPC 2.0 - Review changes.
- fix documentation.
- Add line ending.
- Fix bad usage of string_view

Commit:
JSONRPC 2.0:
- Change autests to work with new traffic_ctl.
- Add new helper function inside trafficserver.ext to help with runroot
  handling.

Commit:
JSONRPC 2.0:
- Change string_view to avoid issues with a bad usage of them
- Add error handling on invalid socket pathname.(unit tests issues)
- Change socket api to better error handling.
- Fix: Use strncpy instead of strcpy

Commit:
JSONRPC - traffic_ctl : Add abs include folder

Commit:
JSONRPC - Adapt rc/* files to work with TS instead of TM

Commit:
JSONRPC - Add JSON Schema for request and response verification.

Commit:
JSONRPC 2.0 - admin_host_set_status, Apply HostStatus API changes.

Commit:
JSONRPC 2.0: IPC server: add support to set the type of permissions(mode) for the jsonrpc socket.

Commit:
JSONRPC: Address some review comments.

Commit:
JSONRPC 2.0: Fix string_view usage. Use std::string instead.

Commit:
JSONRPC 2.0: Move some mgmt stats from TM to TS when stand alone initialization is done for TS. This will help some of the mgmt tools that uses this metrics.

Commit:
JSONRPC - Tidy up the server part.
- Add more unit tests.
- Make error handling a bit clean.
- Log more meaningful error messages.
- Remove the message validation by using a YAML parser and now rely on the  buffer size.

Commit:
JSONRPC 2.0 - docs:
- Add jsonrpc node documentation. API doc now is in his own file.
- Make a better documentation structure.

Commit:
JSONRPC:
- Errata - feed the classification error from a different source

- This commit includes a new test extension which provides a test API to handle jsonrpc requests and responses from the server. Also includes a new Custom JSONRPC response Tester.

- traffic_ctl: Improve output format handling. --format can now take several values.

- unittest:
 Add unit test to validate request and response message using the json schema.
 Extend the jsonrpc extention to accept a schema as part of the AddJsonRPCClientRequest functionality.
 Add new tester JSONRPCResponseSchemaValidator to perform schema validation on jsonrpc responses.

- Doc
  traffic_ctl: Add --format example docs.

bneradt committed:
Revert "Remove incompatible changes for 9.2.0 (#8316)"

This reverts commit 4208ecb8719add76aa46ab2005ff9424531e2e73.

Reverting the removal of the incompatible changes that we took out of
master so that those incompatible changes will be in our 10-Dev branch.

Commit:
[traffic_ctl] - Make traffi_ctl binary to return non-zero code when fail and print to stderr in case of an error. (#560)

Commit:
Fix compiler error. Match fwd with class definition (#566)

bneradt committed:
Fix traffic.out rolling (#569)

traffic.out used to be rotated by traffic_manager. Now that we
transitioned to JSONRPC, traffic_manager no longer runs. This
transitions log rotation handling to traffic_server if JSONRPC is
running.

brbzull0 committed:
Apply inkcore changes and makefile merge fix

brbzull0 committed:
[doc] - jsonrpc, update doc links.

Commit:
[JSONRPC] - Move traffic_ctl JSONRPC request/response definitions to a shared foler (#577)

* [JSONRPC] - Move JSONRPC request/response definitions to a shared folder.
- Split base request/response object into base objects and specific ones for traffic_ctl.
- Add documentation.
- Clean up some code in the RPC client code.

* [JSONRPC] - rpc shared client, add invoke function that takes a JSONRPC object.

* [JSONRPC] - fix typo

* [JSONRPC] - rename namespace. Add default template type.

* Fix typo in yaml_codecs.h

Co-authored-by: Damian Meden <damian.meden@gmail.com>

Commit:
[JSONRPC] - Add update_type to the set_record response in order to let clients be aware of the update logic. This also includes some tidy up on naming as well as making 'traffic_ctl config get' to display errors if they are back from the server. (#580)

Commit:
Remove TM connectivity from traffic_crashlog (#583)

elsloo committed:
Removed references to the throttle option from the slice plugin. (#8373)

elsloo committed:
Added support for promoting internal (plugin-initiated) requests. (#8363)

randall committed:
Adds support for TCP_NOTSENT_LOWAT sockopt (#8354)

Commit:
traffic_top - Make traffic_top work with the jsonrpc node. (#586)

SolidWallOfCode committed:
Doc: Add proxy.config.cacvhe.mutex_retry_delay (#8376)

SolidWallOfCode committed:
Doc: cleanup build errors. (#8377)

Commit:
Fix regex_remap plugin redirection. (#8359)

brbzull0 committed:
[JSONRPC] - Prepare RPC Manager class to handle all sorts of handlers. This commit includes changes in the jsonrpc manager to allow different kind of handlers being implemented and registered, including handlers from plugins(API not included here). The core part of the manager was changed to use a single table for all the handler(previously one per handler type, method and notifications). Also includes a set of tools to allow writing unit tests around the jsonrpc API.

brbzull0 committed:
Remove whitespaces from include/shared/rpc/README.md and remove unused test function.

bneradt committed:
Adding TLS session key logging capability (#8337)

Adding the ability to log TLS session keys to a log file for packet
capture decryption purposes.

This adds the following reloadable configuration:
proxy.config.ssl.keylog_file

Since this can work for QUIC as well, this also deprecates:
proxy.config.quic.client.keylog_file

jrushford committed:
Fixes issue #8329 crash in NextHopConsistentHash where when the only (#8365)

host in a strategy is unavailable due to DNS lookup failures.

- refactored NextHopConsistentHash::findNextHop() to fix the crash
  and simplify it.
- fixed the unit tests in test_NextHopConsistentHash.cc so that test
  failure checks using strcmp() do not crash with a nullptr hostname
  result.

Commit:
Add Au test for pqsi and pqsp log fields. (#8372)

maskit committed:
Fix a potential H2 stall issue (#8381)

dragon512 committed:
body factory does not respect runroot (#8388)

maskit committed:
Update the feature lists available for autest (#8392)

This closes #7075.

hankai17 committed:
Update CacheWrite.cc (#8405)

code optimize for cache_evac debug log

Commit:
Fix map_with_recv_port URL remapping type. (#8406)

ezelkow1 committed:
Add set-body ability to header rewrite (#8411)

ema committed:
Add stats for concurrent stream limits exceeded (#8409)

Introduce two new stats tracking when the limits on maximum concurrent
streams are exceeded for both inbound and outbount streams, respectively
proxy.process.http2.max_concurrent_streams_exceeded_in and
proxy.process.http2.max_concurrent_streams_exceeded_out.

See https://datatracker.ietf.org/doc/html/rfc7540#section-5.1.2

bneradt committed:
AuTest: Execute Test Python Scripts with sys.executable (#8412)

A handful of our AuTests execute their own Python scripts. Generally
these just ran with whatever `python3` picked up. This changes those to
use {sys.executable} which will run those scripts with the same Python
being used for autest itself rather than some other system Python which
may not have the required dependencies installed in it. This way any
requirements for the scripts can be placed in the tests/Pipfile and it
should be available for those scripts.

bryancall committed:
Added required checks before merging (#8413)

bryancall committed:
Don't require branch to be up to date before merging (#8416)

ezelkow1 committed:
Remove operatorbody class, its not needed at this time (#8424)

bneradt committed:
Adding doc/README.md (#8420)

This adds a README describing and providing an example of how to build
the docs.

Fixes #8410

ezelkow1 committed:
add hacktoberfest label (#8425)

randall committed:
Enables crosscompling of ATS (#7786)

These changes allow for CompileParseRules to be built and run on host
architecture along with cleaning up automake macros that previously
prevented crosscompliation

meeramn committed:
Doc: fix typos in Strategy documentation (#8408)

bryancall committed:
Increased the max number of operations for marking PR and issues stale to 100 (#8394)

masaori335 committed:
Add AuTest for stats-over-http plugin (#8422)

zwoop committed:
Make sure the include files are built first (#8427)

Commit:
Rename IpEndpoint member functions named port() to network_order_port(). (#8421)

zwoop committed:
Refactors and promotes the Txn Control mechanism with Get() and Set() (#8428)

* Promote TSHttpTxnCntl to stable ts.h

* Refactor the TSHttpTxnCntl() API into explicit GET/SET

This adds

tsapi bool TSHttpTxnCntlGet(TSHttpTxn txnp, TSHttpCntlType ctrl);
tsapi TSReturnCode TSHttpTxnCntlSet(TSHttpTxn txnp, TSHttpCntlType ctrl, bool data);

* Adds support for the API controlled cacheable flags

This adds

  TS_HTTP_CNTL_RESPONSE_CACHEABLE
  TS_HTTP_CNTL_REQUEST_CACHEABLE
  TS_HTTP_CNTL_SERVER_NO_STORE

* Adds support for the TXN debug flag

This adds

  TS_HTTP_CNTL_TXN_DEBUG

* Adds the control feature for turning off remapping

This adds

  TS_HTTP_CNTL_SKIP_REMAPPING

* Adds docs for new TSHttpTxnCntlGet/Set APIs

* Updates all core plugins to use the new API

randall committed:
tests: Add shbang to python scripts with a main (#8430)

randall committed:
Remove empty tests/unit_tests directoy+makefile (#8429)

randall committed:
Make header_rewrite_cond_cache test version indepdendent (#8431)

randall committed:
Adds new API: TSVConnSslSniGet (#8313)

masaori335 committed:
Fix null-pointer-subtraction warning (#8433)

masaori335 committed:
Fix deprecated-copy warning (#8434)

masaori335 committed:
Fix unused-but-set-variable warnings (#8435)

SolidWallOfCode committed:
Promote class PendingAction from HttpSM.h for use in other classes. (#8423)

randall committed:
rate_limit: convert to using TSVConnSslSniGet (#8414)

jbfavre committed:
Fix traffic_manager build when mime-sanity-check is enabled (#8438)

jbfavre committed:
Fix traffic_top build when using -Werror=format-security (#8437)

bneradt committed:
Update the Multiplexer Docs for Multplexed HTTPS Connections (#8440)

This updates the Multiplexer docs to describe the need for remap.config
entries for the multiplexed hosts. It also describes how these
remap.config rules can be used to establish HTTPS requests to those
hosts.

randall committed:
bigobj: use automake to build test utilities (#8441)

Continuing the work originally done in 61f8ee3bfdea2690677d6ef3697b5aa272ffd297

randall committed:
For verify_global_plugin test, check for platform independent error message (#8442)

This allows the test to work on something other than Linux

Commit:
Make sni.yaml errors cause an unrecoverable TS crash at startup. (#8208)

masaori335 committed:
Fix SSLAddressLookup Test (#8436)

masaori335 committed:
Fix timeout checks of NetHandler::manage_active_queue() (#8287)

randall committed:
Exports all symbols for missing_mangled_definition.so test plugin (#8444)

This fixes the tests verify_global_plugin and verify_remap_plugin when
using lld and LTO

randall committed:
Include CPPFLAGS when building CompileParseRules (#8445)

This is a follow on to 852d90d3

bryancall committed:
Report an error if configure can't find zlib (#8446)

bneradt committed:
Fix Multiplexer POST/PUT Body Handling (#8439)

The previous Multiplexer Plugin test only verified that the plugin would
load without error. This updates the test to verify it can multiplex to
multiple hosts, exercising both HTTP and HTTPS connections to the
multiplexed hosts.

While doing this, I added POST and PUT request method tests. This showed
that the plugin could not handle request bodies and simply failed a
misconfigured assertion when the request body was processed. This change
also fixes that bug in the plugin.

bneradt committed:
Merge remote-tracking branch 'upstream/master' into rebase_10_dev

brbzull0 committed:
JSONRPC:
 - traffic_ctl: config describe/get to validate that the record is registered in ATS before showing any info
 - Protocol: record response - make record/error list always present even if contains no values.
 - Config: Make the jsonrpc socket permissions restricted by default.

brbzull0 committed:
JSONRPC - Fix autest to handle config reload using the new jsonrpc request api.
This change involves removing a hand call to traffic_ctl and using Request.admin_config_reload() instead.
Also adjust the scripts to take the runroot file so it can be passed to traffic_ctl when executed.

brbzull0 committed:
[JSONRPC] - Add new endpoint to fetch the FileManager registry content.

The idea with this change is to expose the content of the bindings inside the FileManager singleton class.
This commit also adds support to query this endpoint using traffic_ctl. Docs are also updated

brbzull0 committed:
JSONRPC - doc. Fix doc indentation

brbzull0 committed:
JSONRPC - Add api doc link as reference for traffic_ctl

brbzull0 committed:
Fix unistd.h include so  global functions gets correctly picked up.
This was making the build fail on FreeBSD.

Commit:
Clean up the shared handler response data. This shared data is used to carry the plugin responses to jsonrpc messages. (#614)

This needs to be clear up after use otherwise the state may hold the old one which can end up giving the wrong response
to rpc requests.

randall committed:
Propagate proxy.config.net.sock_option_flag_in to newly accepted connections (#8623)

* For client connections, disable setting SO_LINGER by default
